src/app/fdctl/run/tiles/fd_repair.c

```diff
@@ -0,0 +1,675 @@
+  ulong *          net_out_sync;
+  ulong            net_out_depth;
+  ulong            net_out_seq;
+
+  fd_wksp_t * net_out_mem;
+  ulong       net_out_chunk0;
+  ulong       net_out_wmark;
+  ulong       net_out_chunk;
+
+  fd_frag_meta_t * store_out_mcache;
+  ulong *          store_out_sync;
+  ulong            store_out_depth;
+  ulong            store_out_seq;
+
+  fd_wksp_t * store_out_mem;
+  ulong       store_out_chunk0;
+  ulong       store_out_wmark;
+  ulong       store_out_chunk;
+
+  uchar src_mac_addr[6];
+
+  /* Includes Ethernet, IP, UDP headers */
+  ulong repair_buffer_sz;
+  uchar repair_buffer[ FD_NET_MTU ];
+};
+typedef struct fd_repair_tile_ctx fd_repair_tile_ctx_t;
+
+static fd_gossip_peer_addr_t *
+resolve_hostport( const char * str /* host:port */, fd_gossip_peer_addr_t * res ) {
+  fd_memset( res, 0, sizeof( fd_gossip_peer_addr_t ) );
+
+  /* Find the : and copy out the host */
+  char buf[128];
+  uint i;
+  for( i = 0;; ++i ) {
+    if( str[i] == '\0' || i > sizeof( buf ) - 1U ) {
+      FD_LOG_ERR( ( "missing colon" ) );
+      return NULL;
+    }
+    if( str[i] == ':' ) {
+      buf[i] = '\0';
+      break;
+    }
+    buf[i] = str[i];
+  }
+  if( i == 0 ) /* :port means $HOST:port */
+    gethostname( buf, sizeof( buf ) );
+
+  struct hostent * host = gethostbyname( buf );
+  if( host == NULL ) {
+    FD_LOG_WARNING( ( "unable to resolve host %s", buf ) );
+    return NULL;
+  }
+  /* Convert result to repair address */
+  res->l    = 0;
+  res->addr = ( (struct in_addr *)host->h_addr )->s_addr;
+  int port  = atoi( str + i + 1 );
+  if( ( port > 0 && port < 1024 ) || port > (int)USHORT_MAX ) {
+    FD_LOG_ERR( ( "invalid port number" ) );
+    return NULL;
+  }
+  res->port = htons( (ushort)port );
+
+  return res;
+}
+
+FD_FN_CONST static inline ulong
+scratch_align( void ) {
+  return 4096UL;
+}
+
+FD_FN_PURE static inline ulong
+loose_footprint( fd_topo_tile_t const * tile FD_PARAM_UNUSED ) {
+  return 1UL * FD_SHMEM_GIGANTIC_PAGE_SZ;
+}
+
+FD_FN_PURE static inline ulong
+scratch_footprint( fd_topo_tile_t const * tile FD_PARAM_UNUSED) {
+
+  ulong l = FD_LAYOUT_INIT;
+  l = FD_LAYOUT_APPEND( l, alignof(fd_repair_tile_ctx_t), sizeof(fd_repair_tile_ctx_t) );
+  l = FD_LAYOUT_APPEND( l, fd_repair_align(), fd_repair_footprint() );
+  l = FD_LAYOUT_APPEND( l, fd_repair_align(), fd_repair_footprint() );
+  l = FD_LAYOUT_APPEND( l, fd_scratch_smem_align(), fd_scratch_smem_footprint( SMAX ) );
+  l = FD_LAYOUT_APPEND( l, fd_scratch_fmem_align(), fd_scratch_fmem_footprint( SDEPTH ) );
+  return FD_LAYOUT_FINI( l, scratch_align() );
+}
+
+FD_FN_CONST static inline void *
+mux_ctx( void * scratch ) {
+  return (void*)fd_ulong_align_up( (ulong)scratch, alignof(fd_repair_tile_ctx_t) );
+}
+typedef struct __attribute__((packed)) {
+  fd_eth_hdr_t eth[1];
+  fd_ip4_hdr_t ip4[1];
+  fd_udp_hdr_t udp[1];
+} eth_ip_udp_t;
+
+static inline void
+populate_packet_header_template( eth_ip_udp_t * pkt,
```

Comment:
Copy pasted from shred? Probably time to move these to a utility location, util/net would be appropriate for this IMO.